include/envoy/network/socket.h

```diff
@@ -222,5 +222,50 @@ using SocketPtr = std::unique_ptr<Socket>;
 using SocketSharedPtr = std::shared_ptr<Socket>;
 using SocketOptRef = absl::optional<std::reference_wrapper<Socket>>;
 
+class SocketInterface {
+public:
+  virtual ~SocketInterface() = default;
+  /**
+   * Low level api to create a socket in the underlying host stack. Does not create an
+   * Envoy socket.
+   * @param type type of socket requested
+   * @param addr_type type of address used with the socket
+   * @param version IP version if address type is IP
+   * @return Socket file descriptor
+   */
+  virtual IoHandlePtr socket(Address::SocketType type, Address::Type addr_type,
+                             Address::IpVersion version) PURE;
+
+  /**
+   * Low level api to create a socket in the underlying host stack. Does not create an
+   * Envoy socket.
+   * @param socket_type type of socket requested
+   * @param addr address that is gleaned for address type and version if needed (@see createSocket)
+   */
+  virtual IoHandlePtr socket(Address::SocketType socket_type,
+                             const Address::InstanceConstSharedPtr addr) PURE;
+
+  /**
+   * Returns true if the given family is supported on this machine.
+   * @param domain the IP family.
+   */
+  virtual bool ipFamilySupported(int domain) PURE;
+
+  /**
+   * Obtain an address from a bound file descriptor. Raises an EnvoyException on failure.
+   * @param fd socket file descriptor
+   * @return InstanceConstSharedPtr for bound address.
+   */
+  virtual Address::InstanceConstSharedPtr addressFromFd(os_fd_t fd) PURE;
```

Comment:
Would it be better to use `IoHandle&` here instead of `os_fd_t`?

Reply:
Eventually we might get away with it (quick search shows that it's always used on IoHandles). But this PR is just a migration from SocketInterface namespace to a virtual class with no functional changes and minimal refactoring. Same applies to peerAddressFromFd.